### PR TITLE
[Feature] Layer-wise MoE auxiliary loss (split finalize) and optional async router D2H offload

### PR DESCRIPTION
**Pull request by tina-wen**
# Summary

Layer-wise split finalize for MoE auxiliary losses and optional async router D2H offload.
## What changed

Implements layer-wise accumulation + finalize for balancing and z-loss `layer_moe_loss.py`.
Adds lazy async D2H offload for router tensors `router_offload.py` and integrates it in `moe.py`.
## Benefits

Reduces peak GPU memory by offloading router logits/weights to pinned CPU while keeping tensor semantics.
Async offload is lazy (waits only when CPU data is actually needed), enabling overlap of D2H copy and GPU work to reduce wall-clock overhead.
 Enabling router_async_offload=True lowers memory at the cost of possible host-device transfer latency.

### CODE REVIEW

File: xtuner/v1/model/moe/moe.py [hunk 1]
```diff
@@ -77,6 +77,22 @@
 MOE_EP_COMPILE_CFG = MOE_NON_EP_COMPILE_CFG.copy()
 MOE_EP_COMPILE_CFG.pop("xtuner.v1.module.decoder_layer.moe_decoder_layer.MoEDecoderLayer.forward")
 
+class _AllReduce(torch.autograd.Function):
```
Comment (pppppM) [inline xtuner/v1/model/moe/moe.py]: Why not use pytorch's interface directly?

https://github.com/pytorch/pytorch/blob/976d4bd88264c4c66c6a6d5c8f69baad3d6bf56e/torch/distributed/nn/functional.py#L205

Comment (tina-wen) [reply xtuner/v1/model/moe/moe.py]: Thanks for the comment. We have replaced the custom implementation with dist.all_reduce as suggested. The issue is resolved.

**Comment by HAOCHENYE** [comment]
@claude review

File: xtuner/v1/loss/aux_loss.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        balancing_ctx: BalancingLossContext | None = None,
+        z_ctx: ZLossContext | None = None,
+        dim: int = 1,
+    ) -> None:
+        """Accumulate routing statistics for one layer."""
+        selected_router_weights, _ = _select_nonpad(router_weights, mask, dim=dim)
+        selected_router_logits, _ = _select_nonpad(router_logits, mask, dim=dim)
+
+        _, selected_experts = torch.topk(selected_router_logits, num_experts_per_tok, dim=-1)
+        tokens_per_expert_logits = torch.histc(
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: This is not logits, renaming it to `tokens_per_expert_per_layer`

File: xtuner/v1/loss/aux_loss.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            selected_experts.view(-1),
+            bins=self.n_routed_experts,
+            min=0,
+            max=self.n_routed_experts,
+        ).to(torch.long)
+
+        if balancing_ctx is not None:
+            balancing_loss_ctx = cast("BalancingLossContext", balancing_ctx)
+            balancing_loss_ctx.local_load_logits_list.append(tokens_per_expert_logits)
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Could we update `local_load_logits_list` in `balancing_loss_ctx.update_split_aux`?

File: xtuner/v1/loss/aux_loss.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        _, selected_experts = torch.topk(selected_router_logits, num_experts_per_tok, dim=-1)
+        tokens_per_expert_logits = torch.histc(
+            selected_experts.view(-1),
+            bins=self.n_routed_experts,
+            min=0,
+            max=self.n_routed_experts,
+        ).to(torch.long)
+
+        if balancing_ctx is not None:
+            balancing_loss_ctx = cast("BalancingLossContext", balancing_ctx)
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Is this `cast` necessary?

File: xtuner/v1/loss/aux_loss.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        if z_ctx is not None:
+            z_loss_ctx = cast("ZLossContext", z_ctx)
+            z_loss_ctx.update_split_aux(
+                router_logits=selected_router_logits,
+            )
+
+    def _cal_tokens_per_expert(self, balancing_ctx: BalancingLossContext | None) -> torch.Tensor:
+        """Get tokens-per-expert tensor for logging/bias update."""
+        if balancing_ctx is not None:
+            local_load_logits = cast("BalancingLossContext", balancing_ctx).local_load_logits_list
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Is this cast necessary?

File: xtuner/v1/loss/aux_loss.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            cast("BalancingLossContext", balancing_ctx).local_load_logits_list = []
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Is this cast necessary?

File: xtuner/v1/loss/aux_loss.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        else:
+            local_load_logits = self._local_load_logits_list
+            self._local_load_logits_list = []
+
+        if not local_load_logits:
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: If `local_load_logits` is an empty list, should we directly raise an error here?

File: xtuner/v1/loss/aux_loss.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        balancing_ctx: BalancingLossContext | None,
+        z_ctx: ZLossContext | None,
+        num_experts_per_tok: int,
+        non_pad_token: int,
+    ) -> tuple[torch.Tensor | None, torch.Tensor | None, torch.Tensor] | None:
+        """Finalize split auxiliary losses and expert counts from runtime
+        state."""
+        balancing_loss = None
+        if balancing_ctx is not None:
+            balancing_loss_ctx = cast("BalancingLossContext", balancing_ctx)
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Is this cast necessary?

File: xtuner/v1/loss/aux_loss.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            balancing_loss = balancing_loss_ctx.finalize_split_aux_loss(
+                n_routed_experts=self.n_routed_experts,
+                num_experts_per_tok=num_experts_per_tok,
+                non_pad_token=non_pad_token,
+            )
+
+        z_loss = None
+        if z_ctx is not None:
+            z_loss_ctx = cast("ZLossContext", z_ctx)
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Is this cast necessary?

File: xtuner/v1/loss/aux_loss.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+            z_loss = z_loss_ctx.finalize_split_aux_loss()
+
+        tokens_per_expert_global = self._cal_tokens_per_expert(balancing_ctx)
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/aux_loss.py]: Since `_local_load_logits_list` will always be updated during `accumulate`, maybe we do not need to pass `balancing_ctx` to `_cal_tokens_per_expert` here.

File: xtuner/v1/loss/moe_loss.py [hunk 11]
```diff
@@ -240,6 +243,84 @@ def forward(
 
         return loss
 
+    def update_split_aux(
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/moe_loss.py]: Could we directly remove `forward` impl?

File: xtuner/v1/loss/moe_loss.py [hunk 12]
```diff
@@ -339,6 +422,55 @@ def forward(self, router_logits: torch.Tensor) -> torch.Tensor:
...
```
Comment (HAOCHENYE) [inline xtuner/v1/loss/moe_loss.py]: Could we directly remove `forward` iml?

**Review by HAOCHENYE — COMMENTED** [review]
Just hold, the train engine refactor will optimizer

File: xtuner/v1/loss/aux_loss.py (hunk 2 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Renamed.

File: xtuner/v1/loss/aux_loss.py (hunk 3 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed `local_load_logits_list` from `balancing_loss_ctx`, replaced it with `local_load`. Now `tokens_per_expert_per_layer` computed in `aux_loss_ctx.accumulate` is directly passed to `balancing_loss_ctx.update_split_aux` to `update local_load`.

File: xtuner/v1/loss/aux_loss.py (hunk 5 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed the unnecessary `cast`.

File: xtuner/v1/loss/aux_loss.py (hunk 4 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed the unnecessary `cast`.

File: xtuner/v1/loss/aux_loss.py (hunk 6 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed the unnecessary `cast`.

File: xtuner/v1/loss/aux_loss.py (hunk 7 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: You're right — an empty list should not happen under normal conditions (that would mean all layers are dense, but then the `ModelConfig` shouldn't inherit from MoE). Fixed by raising `RuntimeError` accordingly.

File: xtuner/v1/loss/aux_loss.py (hunk 8 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed the unnecessary `cast`.

File: xtuner/v1/loss/aux_loss.py (hunk 9 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed the unnecessary `cast`.

File: xtuner/v1/loss/aux_loss.py (hunk 10 above)
Comment (tina-wen) [reply xtuner/v1/loss/aux_loss.py]: Removed `local_load_logits_list` from `balancing_loss_ctx`, replaced it with `local_load`. Now `tokens_per_expert_per_layer` from `aux_loss_ctx.accumulate` is directly passed to `balancing_loss_ctx.update_split_aux` to update `local_load._cal_tokens_per_expert` no longer takes `balancing_ctx` as an argument.

File: xtuner/v1/loss/moe_loss.py (hunk 11 above)
Comment (tina-wen) [reply xtuner/v1/loss/moe_loss.py]: Still used in `qwen3vl_text.py` — added a TODO to remove later.

File: xtuner/v1/loss/moe_loss.py (hunk 12 above)
Comment (tina-wen) [reply xtuner/v1/loss/moe_loss.py]: Still used in `qwen3vl_text.py` — added a TODO to remove later.
